docs/adrs/0275-custom-label-support.md

```diff
@@ -0,0 +1,43 @@
+# ADR 275: Custom Label Support
+**Date**: 2020-01-14
+
+**Status**: Proposed
+
+## Context
+
+In the case of a single repo having multiple self-hosted runners, actions do not currently support any way of selecting which runner or kind of runner a job should run on, other than by architecture or OS.
+
+Example use cases for this could be:
+
+* Your project involves some machine learning models. You want to add a second runner with a GPU for testing your models, and to have your model tests job use this runner, but other jobs to use the runner without a GPU. 
+* You want to use GitHub actions to deploy your app, and for security you use a self-hosted runner inside your isolated network. If you have a separate isolated network for each environment (e.g. testing, staging, production), you would need to run your deploy job on a different runner depending on which environment you are deploying to.
+* Any other situation where you would like to select which self-hosted runner to use for a job based on the capabilities or characteristics of the runner.
+
+GitHub actions already support selecting a job runner based on labels, and automatically give runners labels based on OS and arch, so a natural extension would be to allow operators of self-hosted runners to add their own labels, so that users can select job runners based on any operator-defined characteristic.
+
+See Issue: https://github.com/actions/runner/issues/262
+
+## Decision
+
+This ADR proposes that we add a `--label` option to `config.sh`, which could be used to add custom additional labels to the configured runner.
+
+For example, to add a single extra label the operator could run:
+```bash
+./config.sh --label my-extra-label
+```
+Which would add the label `my-extra-label` to the runner, and enable users to select the runner in their workflow using this label:
+```yaml
+runs-on: [self-hosted, my-extra-label]
+```
+
+To add multiple labels:
+```bash
+./config.sh --label my-extra-label --label yet-another-label
+```
+This would add both labels `my-extra-label` and `yet-another-label`.
+
+## Consequences
+
+The ability to add custom labels to a self-hosted runner would enable most scenarios where job runner selection based on runner capabilities or characteristics are required.
+
+There could be unexpected behavior if an operator adds inappropriate labels, e.g. they add the label `linux` to a Windows runner, but this would be considered operator error.
```

Comment:
We should consider not allowing the user to add the well-known labels to a runner as a custom label. I don't think users will intentionally do this, but in the case of copy-paste error or a user running the wrong script will result in some confusion.

Reply:
There could be situations where it is valid to add these well-known labels - for example: you have a runner on x86 system, but the system has binfmt setup to allow seamless execution of arm binaries via qemu, so that you can run unit tests on applications built for resource-constrained arm devices like the raspberry pi, where running the tests on actual hardware would be very slow. In this case the system is capable of executing both x86 and arm binaries. You can have a similar setup in windows subsystem for linux, where both windows and linux binaries are executable.

Reply:
r2zero is correct, this was an explicit goal of the initial label support.  An org level admin to explicitly overriding the use of a hosted pool in his org was a scenario.  @chrispat

Reply:
That makes sense to me  👍